clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4225,76 +4232,84 @@ public ListOffsetsResult listOffsets(Map<TopicPartition, OffsetSpec> topicPartit
             }
         }
 
-        for (final Map.Entry<Node, Map<String, ListOffsetsTopic>> entry : leaders.entrySet()) {
-            final int brokerId = entry.getKey().id();
+        for (final Map.Entry<Node, Map<ListOffsetRequestVersion, Map<String, ListOffsetsTopic>>> versionedEntry : leaders.entrySet()) {
+            for (final Map.Entry<ListOffsetRequestVersion, Map<String, ListOffsetsTopic>> entry : versionedEntry.getValue().entrySet()) {
+                final int brokerId = versionedEntry.getKey().id();
 
-            calls.add(new Call("listOffsets on broker " + brokerId, context.deadline(), new ConstantNodeIdProvider(brokerId)) {
+                calls.add(new Call("listOffsets on broker " + brokerId, context.deadline(), new ConstantNodeIdProvider(brokerId)) {
 
-                final List<ListOffsetsTopic> partitionsToQuery = new ArrayList<>(entry.getValue().values());
+                    final List<ListOffsetsTopic> partitionsToQuery = new ArrayList<>(entry.getValue().values());
 
-                @Override
-                ListOffsetsRequest.Builder createRequest(int timeoutMs) {
-                    return ListOffsetsRequest.Builder
+                    @Override
+                    ListOffsetsRequest.Builder createRequest(int timeoutMs) {
+                        ListOffsetRequestVersion requestVersion = entry.getKey();
+                        if (requestVersion == ListOffsetRequestVersion.V7AndAbove) {
+                            return ListOffsetsRequest.Builder
+                                .forMaxTimestamp(context.options().isolationLevel())
+                                .setTargetTimes(partitionsToQuery);
+                        }
```

Comment:
I'd like to better understand how we handle a broker which would not support the version that we need. 

`ListOffsetsRequest.Builder.forMaxTimestamp` constrains the version to 7 and above when we have have at least one max timestamp spec. It the broker does not support version 7, the request is failed with an `UnsupportedVersionException` and we fail all the future of the brokers with it in `handleFailure`.

Now, let's imagine a case where the user does not only include "max timestamp specs" in his request. At the moment, we fail all of them irrespectively of their type. I wonder if we should retry to other specs in this particular case. Have we considered doing this?

Reply:
At present the only way requestVersion could be V7AndAbove is if we were issuing MAX_TIMESTAMP requests because of the way the calls are parsed earlier:

`                    
ListOffsetRequestVersion requiredRequestVersion = offsetQuery == ListOffsetsRequest.MAX_TIMESTAMP
                        ? ListOffsetRequestVersion.V7AndAbove :
                        ListOffsetRequestVersion.V0AndAbove;
`

All non-max timestamp requests are built using forConsumer rather than forMaxTimestamp and so should succeed against older brokers. Maybe the enums are a bit misleading in this regard. I'll see if i can come up with something better.

Reply:
Oh... I see. So you are saying that if we have two specs for a given leader, say one with MAX_TIMESTAMP and another one with EARLIEST_TIMESTAMP, we send two separate requests to that leader, right?

Reply:
That's right, as I read it we would send separate requests even under the old logic (i.e. for LATEST_TIMESTAMP and EARLIEST_TIMESTAMP). The only difference here is we limit the versions for MAX_TIMESTAMP.

Reply:
It seems that the current logic send only one request per broker/leader whereas we could send up to two requests with your PR because specs are partitioned by `Node` and `ListOffsetRequestVersion`. Previously, they were only partitioned by `Node`.

Intuitively, I would have approached the problem differently. I would have put all the specs in the same request and constrained its version to 7 and above if there is at least one `MAX_TIMESTAMP`. If the request succeeds, all good. If the request fail with an `UnsupportedVersionException`, I would have retried with all the specs but the `MAX_TIMESTAMP` ones and I would have failed the future of the `MAX_TIMESTAMP` specs.

In case of an `UnsupportedVersionException`, the admin client calls the `handleUnsupportedVersionException` method of the `Call`. This gives you an opportunity to downgrade and to retry the `Call`. There are couple of example in the `KafkaAdminClient`.

I wonder if we could rely on a similar pattern and avoid sending two requests per leader in the worst case. What do you think?

Reply:
Ah, I missed that pattern. I'll fix it up to be consistent with the other examples.

Reply:
@dajac I changed to use the approach discussed here, I'd appreciate it if you can give this another review.